File: Toggly.FeatureManagement.Android/toggly-android-core/src/main/kotlin/io/toggly/core/SignedDefsVerify.kt
package io.toggly.core

import io.toggly.core.models.FeatureFlags
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.Response
import java.math.BigInteger
import java.security.AlgorithmParameters
import java.security.KeyFactory
import java.security.MessageDigest
import java.security.Signature
import java.security.interfaces.ECPublicKey
import java.security.spec.ECGenParameterSpec
import java.security.spec.ECParameterSpec
import java.security.spec.ECPoint
import java.security.spec.ECPublicKeySpec

/**
 * Verifies signed definitions without altering the signed JSON bytes.
 *
 * The definitions service signs the exact raw defs JSON plus its timestamp,
 * using a double SHA-256 digest and an ES256 P-256 signature.
 */
internal object SignedDefsVerify {
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    internal data class SignedEnvelope(
        val defsRaw: String,
        val signature: String,
        val timestamp: Long,
        val kid: String
    )

    /**
     * Extracts the exact raw JSON text of a top-level property.
     */
    internal fun extractRawJsonProperty(text: String, key: String): String? {
        var index = 0
        var depth = 0
        var inString = false
        var escape = false

        while (index < text.length) {
            val character = text[index]
            if (inString) {
                if (escape) {
                    escape = false
                } else if (character == '\\') {
                    escape = true
                } else if (character == '"') {
                    inString = false
                }
                index++
                continue
            }

            when (character) {
                '"' -> {
                    if (depth == 1) {
                        val keyEnd = findStringEnd(text, index) ?: return null
                        val propertyName = text.substring(index + 1, keyEnd)
                        var valueStart = keyEnd + 1
                        while (valueStart < text.length && text[valueStart].isWhitespace()) {
                            valueStart++
                        }
                        if (propertyName == key && valueStart < text.length && text[valueStart] == ':') {
                            valueStart++
                            while (valueStart < text.length && text[valueStart].isWhitespace()) {
                                valueStart++
                            }
                            return extractJsonValue(text, valueStart)
                        }
                        index = keyEnd + 1
                        continue
                    }
                    inString = true
                }

                '{', '[' -> depth++
                '}', ']' -> depth--
            }
            index++
        }

        return null
    }

    internal fun parseSignedEnvelope(body: String): SignedEnvelope {
        val envelope = json.parseToJsonElement(body).jsonObject
        val defsRaw = extractRawJsonProperty(body, "defs")
            ?: extractRawJsonProperty(body, "data")
            ?: throw IllegalArgumentException("Signed envelope missing defs")

        return SignedEnvelope(
            defsRaw = defsRaw,
            signature = envelope.string("signature").also {
                require(it.isNotEmpty()) { "Missing signature" }
            },
            timestamp = envelope.string("timestamp").toLongOrNull()
                ?: throw IllegalArgumentException("Invalid signed timestamp"),
            kid = envelope.string("kid").also {
                require(it.isNotEmpty()) { "Missing kid" }
            }
        )
    }

    internal fun parseDefinitions(defsRaw: String): FeatureFlags {
        return json.decodeFromString(defsRaw)
    }

    internal fun doubleSha256(payload: String): ByteArray {
        val firstDigest = MessageDigest.getInstance("SHA-256")
            .digest(payload.toByteArray(Charsets.UTF_8))
        return MessageDigest.getInstance("SHA-256").digest(firstDigest)
    }

    internal fun verify(
        envelope: SignedEnvelope,
        jwksBody: String
    ) {
        val keys = json.parseToJsonElement(jwksBody).jsonObject["keys"]?.jsonArray
            ?: throw IllegalArgumentException("JWKS missing keys")
        val jwk = findJwk(keys.map { it.jsonObject }, envelope.kid)
            ?: throw IllegalArgumentException("No matching JWK for kid ${envelope.kid}")
        val key = createPublicKey(jwk, envelope.kid)
        val digest = doubleSha256("${envelope.defsRaw}|${envelope.timestamp}")
        val signature = decodeBase64(envelope.signature)
        val derSignature = if (signature.size == 64) p1363ToDer(signature) else signature

        val verifier = Signature.getInstance("NONEwithECDSA")
        verifier.initVerify(key)
        verifier.update(digest)
        if (!verifier.verify(derSignature)) {
            throw IllegalArgumentException("Invalid signature")
        }
    }

    internal fun fetchJwks(response: Response): String {
        if (!response.isSuccessful) {
            throw TogglyException.HttpError(response.code, response.message)
        }
        return response.body?.string()
            ?: throw TogglyException.InvalidResponse("Empty JWKS response body")
    }

    private fun findJwk(keys: List<JsonObject>, kid: String): JsonObject? {
        for (key in keys) {
            if (key["kid"]?.jsonPrimitive?.content == kid) {
                return key
            }
        }
        return null
    }

    private fun findStringEnd(text: String, start: Int): Int? {
        var escape = false
        for (index in start + 1 until text.length) {
            when {
                escape -> escape = false
                text[index] == '\\' -> escape = true
                text[index] == '"' -> return index
            }
        }
        return null
    }

    private fun extractJsonValue(text: String, start: Int): String? {
        if (start >= text.length) {
            return null
        }

        return when (text[start]) {
            '{', '[' -> extractCompositeValue(text, start)
            '"' -> {
                val end = findStringEnd(text, start) ?: return null
                text.substring(start, end + 1)
            }

            else -> {
                var end = start
                while (end < text.length && text[end] !in ",}] \t\r\n") {
                    end++
                }
                text.substring(start, end).takeIf { it.isNotEmpty() }
            }
        }
    }

    private fun extractCompositeValue(text: String, start: Int): String? {
        var depth = 0
        var inString = false
        var escape = false
        for (index in start until text.length) {
            val character = text[index]
            if (inString) {
                if (escape) {
                    escape = false
                } else if (character == '\\') {
                    escape = true
                } else if (character == '"') {
                    inString = false
                }
                continue
            }

            when (character) {
                '"' -> inString = true
                '{', '[' -> depth++
                '}', ']' -> {
                    depth--
                    if (depth == 0) {
                        return text.substring(start, index + 1)
                    }
                }
            }
        }
        return null
    }

    private fun createPublicKey(jwk: JsonObject, expectedKid: String): ECPublicKey {
        require((jwk["kty"]?.jsonPrimitive?.content ?: "EC") == "EC") { "Unsupported JWK type" }
        require((jwk["alg"]?.jsonPrimitive?.content ?: "ES256") == "ES256") { "Unsupported JWK algorithm" }
        require((jwk["crv"]?.jsonPrimitive?.content ?: "P-256") == "P-256") { "Unsupported JWK curve" }

        val x = decodeBase64(jwk.string("x"))
        val y = decodeBase64(jwk.string("y"))
        val calculatedKid = calculateKid(x, y)
        require(jwk.string("kid") == calculatedKid) { "JWK kid does not match coordinates" }
        require(calculatedKid == expectedKid) { "Signed response kid does not match JWK" }

        val parameters = AlgorithmParameters.getInstance("EC").apply {
            init(ECGenParameterSpec("secp256r1"))
        }.getParameterSpec(ECParameterSpec::class.java)
        val keySpec = ECPublicKeySpec(
            ECPoint(BigInteger(1, x), BigInteger(1, y)),
            parameters
        )
        return KeyFactory.getInstance("EC").generatePublic(keySpec) as ECPublicKey
    }

    private fun JsonObject.string(name: String): String {
        return this[name]?.jsonPrimitive?.content
            ?: throw IllegalArgumentException("Missing $name")
    }

    private fun calculateKid(x: ByteArray, y: ByteArray): String {
        val coordinates = ByteArray(x.size + y.size)
        x.copyInto(coordinates)
        y.copyInto(coordinates, destinationOffset = x.size)
        val sha1 = MessageDigest.getInstance("SHA-1").digest(coordinates)
        return sha1.joinToString(separator = "") { "%02X".format(it) } + "ES256"
    }

    /**
     * Decode standard or URL-safe Base64 without java.util.Base64 (API 26+)
     * or android.util.Base64 (unavailable in plain JVM unit tests).
     */
    private fun decodeBase64(value: String): ByteArray {
        val normalized = value.replace('-', '+').replace('_', '/')
        val padded = normalized.padEnd((normalized.length + 3) / 4 * 4, '=')
        require(padded.length % 4 == 0) { "Invalid base64 length" }

        val out = ArrayList<Byte>(padded.length / 4 * 3)
        var index = 0
        while (index < padded.length) {
            val n0 = base64AlphabetValue(padded[index++])
            val n1 = base64AlphabetValue(padded[index++])
            val c2 = padded[index++]
            val c3 = padded[index++]
            val n2 = if (c2 == '=') 0 else base64AlphabetValue(c2)
            val n3 = if (c3 == '=') 0 else base64AlphabetValue(c3)
            val triple = (n0 shl 18) or (n1 shl 12) or (n2 shl 6) or n3
            out.add((triple shr 16).toByte())
            if (c2 != '=') {
                out.add((triple shr 8).toByte())
            }
            if (c3 != '=') {
                out.add(triple.toByte())
            }
        }
        return out.toByteArray()
    }

    private fun base64AlphabetValue(character: Char): Int {
        return when (character) {
            in 'A'..'Z' -> character - 'A'
            in 'a'..'z' -> character - 'a' + 26
            in '0'..'9' -> character - '0' + 52
            '+' -> 62
            '/' -> 63
            else -> throw IllegalArgumentException("Invalid base64 character: $character")
        }
    }

    private fun p1363ToDer(signature: ByteArray): ByteArray {
        require(signature.size == 64) { "P1363 ES256 signatures must be 64 bytes" }
        val r = derInteger(signature.copyOfRange(0, 32))
        val s = derInteger(signature.copyOfRange(32, 64))
        val sequenceLength = r.size + s.size
        require(sequenceLength < 128) { "Unsupported ECDSA signature length" }
        return byteArrayOf(0x30, sequenceLength.toByte()) + r + s
    }

    private fun derInteger(value: ByteArray): ByteArray {
        val firstNonZero = value.indexOfFirst { it != 0.toByte() }
        val trimmed = if (firstNonZero == -1) byteArrayOf(0) else value.copyOfRange(firstNonZero, value.size)
        val encoded = if (trimmed[0].toInt() and 0x80 != 0) byteArrayOf(0) + trimmed else trimmed
        return byteArrayOf(0x02, encoded.size.toByte()) + encoded
    }
}

File: Toggly.FeatureManagement.iOS/TogglyCore/Sources/Services/SignedDefsVerify.swift
import CryptoKit
import Foundation
import Security

/// Errors raised while verifying signed feature definitions.
public enum SignedDefsVerifyError: LocalizedError, Equatable {
    case invalidEnvelope
    case missingDefs
    case kidNotAllowed(String)
    case noMatchingJwk(String)
    case unsupportedAlg(String)
    case unsupportedCurve(String)
    case missingCoordinates
    case invalidKid(expected: String, got: String)
    case invalidSignature
    case invalidKey
    case invalidBase64

    public var errorDescription: String? {
        switch self {
        case .invalidEnvelope:
            return "Invalid signed definitions envelope"
        case .missingDefs:
            return "Signed envelope missing defs"
        case .kidNotAllowed(let kid):
            return "kid not allowed: \(kid)"
        case .noMatchingJwk(let kid):
            return "no matching jwk for kid \"\(kid)\""
        case .unsupportedAlg(let alg):
            return "unsupported alg: \(alg)"
        case .unsupportedCurve(let crv):
            return "unsupported crv: \(crv)"
        case .missingCoordinates:
            return "missing x or y coordinate"
        case .invalidKid(let expected, let got):
            return "invalid kid: expected \(expected), got \(got)"
        case .invalidSignature:
            return "invalid signature"
        case .invalidKey:
            return "invalid verification key"
        case .invalidBase64:
            return "invalid base64"
        }
    }
}

/// Signed definitions response envelope.
public struct SignedEnvelope: Codable, Sendable {
    public let defs: FeatureFlags?
    public let data: FeatureFlags?
    public let signature: String
    public let timestamp: Int64
    public let kid: String

    public init(
        defs: FeatureFlags? = nil,
        data: FeatureFlags? = nil,
        signature: String,
        timestamp: Int64,
        kid: String
    ) {
        self.defs = defs
        self.data = data
        self.signature = signature
        self.timestamp = timestamp
        self.kid = kid
    }
}

/// JSON Web Key used for signed definitions verification.
public struct Jwk: Codable, Sendable {
    public let kty: String?
    public let use: String?
    public let kid: String
    public let crv: String?
    public let x: String?
    public let y: String?
    public let alg: String?

    public init(
        kty: String? = "EC",
        use: String? = "sig",
        kid: String,
        crv: String? = "P-256",
        x: String?,
        y: String?,
        alg: String? = "ES256"
    ) {
        self.kty = kty
        self.use = use
        self.kid = kid
        self.crv = crv
        self.x = x
        self.y = y
        self.alg = alg
    }
}

/// JWKS document from `/.well-known/jwks`.
public struct JwkSet: Codable, Sendable {
    public let keys: [Jwk]

    public init(keys: [Jwk]) {
        self.keys = keys
    }
}

/// Production-compatible signed definitions verification (ES256 / P-256).
///
/// Matches Go / Node / Flutter / Android:
/// - payload = exact raw defs JSON + "|" + timestamp
/// - digest  = SHA-256(SHA-256(utf8(payload)))
/// - signature = standard or URL-safe base64 of IEEE P1363 (r||s), with DER fallback
enum SignedDefsVerify {
    /// Extract the exact raw JSON text of a **top-level** property only.
    /// Nested keys (e.g. `data.defs`) are ignored so unsigned outer fields cannot
    /// be swapped in after verifying nested signed bytes.
    static func extractRawJsonProperty(from text: String, key: String) -> String? {
        var index = text.startIndex
        var depth = 0
        var inString = false
        var escape = false

        while index < text.endIndex {
            let character = text[index]
            if inString {
                if escape {
                    escape = false
                } else if character == "\\" {
                    escape = true
                } else if character == "\"" {
                    inString = false
                }
                index = text.index(after: index)
                continue
            }

            if character == "\"" {
                if depth == 1 {
                    guard let keyEnd = findStringEnd(in: text, startingAt: index) else {
                        return nil
                    }
                    let nameStart = text.index(after: index)
                    let propertyName = String(text[nameStart..<keyEnd])
                    var valueStart = text.index(after: keyEnd)
                    while valueStart < text.endIndex, text[valueStart].isWhitespace {
                        valueStart = text.index(after: valueStart)
                    }
                    if propertyName == key, valueStart < text.endIndex, text[valueStart] == ":" {
                        valueStart = text.index(after: valueStart)
                        while valueStart < text.endIndex, text[valueStart].isWhitespace {
                            valueStart = text.index(after: valueStart)
                        }
                        return extractJsonValue(from: text, startingAt: valueStart)
                    }
                    index = text.index(after: keyEnd)
                    continue
                }
                inString = true
                index = text.index(after: index)
                continue
            }

            if character == "{" || character == "[" {
                depth += 1
            } else if character == "}" || character == "]" {
                depth -= 1
            }
            index = text.index(after: index)
        }

        return nil
    }

    /// Parse verified defs JSON bytes into feature flags (do not use envelope fields).
    static func parseDefinitions(_ defsRaw: String) throws -> FeatureFlags {
        guard let data = defsRaw.data(using: .utf8) else {
            throw SignedDefsVerifyError.invalidEnvelope
        }
        do {
            return try JSONDecoder().decode(FeatureFlags.self, from: data)
        } catch {
            throw SignedDefsVerifyError.invalidEnvelope
        }
    }

    /// Parse a signed envelope and return the exact raw defs (or data) JSON substring.
    static func parseSignedEnvelope(_ bodyText: String) throws -> (envelope: SignedEnvelope, defsRaw: String) {
        guard let data = bodyText.data(using: .utf8) else {
            throw SignedDefsVerifyError.invalidEnvelope
        }

        let envelope: SignedEnvelope
        do {
            envelope = try JSONDecoder().decode(SignedEnvelope.self, from: data)
        } catch {
            throw SignedDefsVerifyError.invalidEnvelope
        }

        guard !envelope.signature.isEmpty, !envelope.kid.isEmpty else {
            throw SignedDefsVerifyError.invalidEnvelope
        }

        guard
            let defsRaw = extractRawJsonProperty(from: bodyText, key: "defs")
                ?? extractRawJsonProperty(from: bodyText, key: "data")
        else {
            throw SignedDefsVerifyError.missingDefs
        }

        return (envelope, defsRaw)
    }

    /// Compute kid = SHA1(x_bytes || y_bytes).hexUpper + "ES256".
    static func computeKid(x: String, y: String) throws -> String {
        let xBytes = try base64ToData(x)
        let yBytes = try base64ToData(y)
        var combined = Data()
        combined.append(xBytes)
        combined.append(yBytes)
        let digest = Insecure.SHA1.hash(data: combined)
        let hex = digest.map { String(format: "%02X", $0) }.joined()
        return "\(hex)ES256"
    }

    /// Verify a signed definitions envelope using exact raw defs bytes.
    static func verifySignedDefinitions(
        defsRaw: String,
        signature: String,
        timestamp: Int64,
        kid: String,
        jwks: JwkSet,
        allowedKids: [String]? = nil
    ) throws {
        if let allowedKids, !allowedKids.isEmpty, !allowedKids.contains(kid) {
            throw SignedDefsVerifyError.kidNotAllowed(kid)
        }

        guard let matching = jwks.keys.first(where: { $0.kid == kid }) else {
            throw SignedDefsVerifyError.noMatchingJwk(kid)
        }

        if let alg = matching.alg, alg != "ES256" {
            throw SignedDefsVerifyError.unsupportedAlg(alg)
        }
        if let crv = matching.crv, crv != "P-256" {
            throw SignedDefsVerifyError.unsupportedCurve(crv)
        }
        guard let x = matching.x, let y = matching.y else {
            throw SignedDefsVerifyError.missingCoordinates
        }

        let expectedKid = try computeKid(x: x, y: y)
        if matching.kid != expectedKid {
            throw SignedDefsVerifyError.invalidKid(expected: expectedKid, got: matching.kid)
        }

        let payload = "\(defsRaw)|\(timestamp)"
        guard let payloadData = payload.data(using: .utf8) else {
            throw SignedDefsVerifyError.invalidEnvelope
        }

        let firstDigest = Data(SHA256.hash(data: payloadData))
        let doubleDigest = Data(SHA256.hash(data: firstDigest))
        let signatureData = try base64ToData(signature)
        let publicKey = try createPublicKey(x: x, y: y)

        let derSignature: Data
        if signatureData.count == 64 {
            guard let converted = p1363ToDER(signatureData) else {
                throw SignedDefsVerifyError.invalidSignature
            }
            derSignature = converted
        } else {
            derSignature = signatureData
        }

        var error: Unmanaged<CFError>?
        let ok = SecKeyVerifySignature(
            publicKey,
            .ecdsaSignatureDigestX962SHA256,
            doubleDigest as CFData,
            derSignature as CFData,
            &error
        )

        if !ok {
            throw SignedDefsVerifyError.invalidSignature
        }
    }

    // MARK: - Helpers

    private static func findStringEnd(in text: String, startingAt startQuote: String.Index) -> String.Index? {
        var escape = false
        var index = text.index(after: startQuote)
        while index < text.endIndex {
            let character = text[index]
            if escape {
                escape = false
            } else if character == "\\" {
                escape = true
            } else if character == "\"" {
                return index
            }
            index = text.index(after: index)
        }
        return nil
    }

    private static func extractJsonValue(from text: String, startingAt start: String.Index) -> String? {
        guard start < text.endIndex else {
            return nil
        }

        let first = text[start]
        if first == "{" || first == "[" {
            var depth = 0
            var inString = false
            var escape = false
            var index = start
            while index < text.endIndex {
                let character = text[index]
                if inString {
                    if escape {
                        escape = false
                    } else if character == "\\" {
                        escape = true
                    } else if character == "\"" {
                        inString = false
                    }
                } else if character == "\"" {
                    inString = true
                } else if character == "{" || character == "[" {
                    depth += 1
                } else if character == "}" || character == "]" {
                    depth -= 1
                    if depth == 0 {
                        return String(text[start...index])
                    }
                }
                index = text.index(after: index)
            }
            return nil
        }

        if first == "\"" {
            guard let end = findStringEnd(in: text, startingAt: start) else {
                return nil
            }
            return String(text[start...end])
        }

        var end = start
        while end < text.endIndex {
            let character = text[end]
            if character.isWhitespace || character == "," || character == "}" || character == "]" {
                break
            }
            end = text.index(after: end)
        }
        let value = String(text[start..<end])
        return value.isEmpty ? nil : value
    }

    static func base64ToData(_ value: String) throws -> Data {
        let normalized = value
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        let remainder = normalized.count % 4
        let padded: String
        if remainder == 0 {
            padded = normalized
        } else {
            padded = normalized + String(repeating: "=", count: 4 - remainder)
        }
        guard let data = Data(base64Encoded: padded) else {
            throw SignedDefsVerifyError.invalidBase64
        }
        return data
    }

    static func base64URLEncode(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    private static func createPublicKey(x: String, y: String) throws -> SecKey {
        var xBytes = try base64ToData(x)
        var yBytes = try base64ToData(y)
        xBytes = padLeft(xBytes, to: 32)
        yBytes = padLeft(yBytes, to: 32)

        var x963 = Data([0x04])
        x963.append(xBytes)
        x963.append(yBytes)

        let attributes: [String: Any] = [
            kSecAttrKeyType as String: kSecAttrKeyTypeECSECPrimeRandom,
            kSecAttrKeyClass as String: kSecAttrKeyClassPublic,
            kSecAttrKeySizeInBits as String: 256
        ]

        var error: Unmanaged<CFError>?
        guard let key = SecKeyCreateWithData(x963 as CFData, attributes as CFDictionary, &error) else {
            throw SignedDefsVerifyError.invalidKey
        }
        return key
    }

    private static func padLeft(_ data: Data, to length: Int) -> Data {
        if data.count >= length {
            return data.suffix(length)
        }
        return Data(repeating: 0, count: length - data.count) + data
    }

    /// Convert IEEE P1363 (r||s) to X9.62 / DER ECDSA signature.
    static func p1363ToDER(_ signature: Data) -> Data? {
        guard signature.count == 64 else { return nil }
        let r = encodeASN1Integer(signature.prefix(32))
        let s = encodeASN1Integer(signature.suffix(32))

        var sequence = Data()
        sequence.append(0x30)
        let contentLength = r.count + s.count
        sequence.append(contentsOf: encodeLength(contentLength))
        sequence.append(r)
        sequence.append(s)
        return sequence
    }

    /// Convert DER ECDSA signature to IEEE P1363 (r||s).
    static func derToP1363(_ der: Data) -> Data? {
        var index = der.startIndex
        guard index < der.endIndex, der[index] == 0x30 else { return nil }
        index = der.index(after: index)

        guard let (_, afterLength) = readLength(der, at: index) else { return nil }
        index = afterLength

        guard let (r, afterR) = readASN1Integer(der, at: index) else { return nil }
        index = afterR
        guard let (s, _) = readASN1Integer(der, at: index) else { return nil }

        var result = Data()
        result.append(padLeft(r, to: 32))
        result.append(padLeft(s, to: 32))
        return result
    }

    private static func encodeASN1Integer(_ bytes: Data.SubSequence) -> Data {
        var value = Data(bytes)
        while value.count > 1, value[0] == 0x00, value[1] & 0x80 == 0 {
            value.removeFirst()
        }
        if value.isEmpty {
            value = Data([0x00])
        }
        if value[0] & 0x80 != 0 {
            value.insert(0x00, at: 0)
        }

        var encoded = Data([0x02])
        encoded.append(contentsOf: encodeLength(value.count))
        encoded.append(value)
        return encoded
    }

    private static func encodeLength(_ length: Int) -> [UInt8] {
        if length < 0x80 {
            return [UInt8(length)]
        }
        if length < 0x100 {
            return [0x81, UInt8(length)]
        }
        return [0x82, UInt8((length >> 8) & 0xff), UInt8(length & 0xff)]
    }

    private static func readLength(_ data: Data, at index: Data.Index) -> (Int, Data.Index)? {
        guard index < data.endIndex else { return nil }
        let first = data[index]
        var next = data.index(after: index)
        if first & 0x80 == 0 {
            return (Int(first), next)
        }
        let count = Int(first & 0x7f)
        guard count > 0, data.distance(from: next, to: data.endIndex) >= count else { return nil }
        var length = 0
        for _ in 0..<count {
            length = (length << 8) | Int(data[next])
            next = data.index(after: next)
        }
        return (length, next)
    }

    private static func readASN1Integer(_ data: Data, at index: Data.Index) -> (Data, Data.Index)? {
        guard index < data.endIndex, data[index] == 0x02 else { return nil }
        let afterTag = data.index(after: index)
        guard let (length, afterLength) = readLength(data, at: afterTag) else { return nil }
        guard data.distance(from: afterLength, to: data.endIndex) >= length else { return nil }
        let end = data.index(afterLength, offsetBy: length)
        var value = data.subdata(in: afterLength..<end)
        while value.count > 1, value[0] == 0x00 {
            value.removeFirst()
        }
        return (value, end)
    }
}
